### PR TITLE
update paper 1.20.6

### PR DESCRIPTION
**Pull request by ColdeZhang**
Try to update for 1.20.6.

Known problems:

 - [x] `kyori:adventure` versions doesnt match, and `4.17.0` is not available; 
 - [x] `ChunkStatus.java` class missing;
 - [x] Some more typing error happened by migration;

### CODE REVIEW

**Comment by MiniDigger** [comment]
https://github.com/KyoriPowered/adventure/releases/tag/v4.17.0 was just released

**Comment by Luuzzi** [comment]
btw

remapper - net.fabricmc:tiny-remapper:0.10.1:fat -> net.fabricmc:tiny-remapper:0.10.2:fat
decompiler - net.minecraftforge:forgeflower:2.0.627.2 -> org.vineflower:vineflower:1.10.1

I think the changes are minor, but it still seems necessary

**Comment by ColdeZhang** [comment]
> btw
> 
> remapper - net.fabricmc:tiny-remapper:0.10.1:fat -> net.fabricmc:tiny-remapper:0.10.2:fat decompiler - net.minecraftforge:forgeflower:2.0.627.2 -> org.vineflower:vineflower:1.10.1
> 
> I think the changes are minor, but it still seems necessary

Yes, UR right. It vital for some class to be decompiled.

**Comment by ColdeZhang** [comment]
> > https://github.com/KyoriPowered/adventure/releases/tag/v4.17.0 was just released
> 
> I found it. But upstream paper using `4.17.0-snapshot` in api, and `4.16.0` in server. I noticed in server build script have a commnet saying "Keep in sync with adventureVersion from Paper-API build file".
> 
> So I'm not sure if it should be modified in folia patch, I think maybe its belonging to paper's patch?

This has been fixed by [Paper](https://github.com/PaperMC/Paper/commit/5729b294eef2ffde690f73c5b14a64a31f301084)

**Comment by ColdeZhang** [comment]
Migrations from folia's patch are done, there still a syntax err in `net/minecraft/core/dispenser/ProjectileDispenseBehavior.java` line 48 `DispenserBlock.eventFired` should use `get()` to fetch bollen type. I leave it alone, because it doesnt changed in folias patch, I think it's upstream's work.

Anyway, patches can be compiled allreday, but I found server stucked when preparing level (I think I may need some help).

**Comment by ColdeZhang** [comment]
Server crashed when prepareing level...
[crash-2024-05-09_15.09.31-server.txt](https://github.com/PaperMC/Folia/files/15259029/crash-2024-05-09_15.09.31-server.txt)

**Comment by ColdeZhang** [comment]
image
 
Greetings, maintainers. I'm done with 1.20.6 update. Could you review it please?

**Comment by ColdeZhang** [comment]
Fine, I got new problem.

Any commands send by client will cause "Internal Exception: java.lang.UnsupportedOperationException", and then kicked.
Commands send my console is ok.

This fault has no stacktrace, just "Internal Exception", I dont know how to debug it.

**Comment by electronicboy** [comment]
enable debug in server.properties or attach a profiler which can do exception recording

but, large PRs should generally be cleared by leaf first, not sure how much acceptance there is towards pulling in an update from somebody else

**Comment by Luuzzi** [comment]
> enable debug in server.properties or attach a profiler which can do exception recording
> 
> but, large PRs should generally be cleared by leaf first, not sure how much acceptance there is towards pulling in an update from somebody else

Is he on vacation now? There is no information when we will be able to get a review of this PR or update to 1.20.6?

**Comment by PedroMPagani** [comment]
> Fine, I got new problem.
> 
> Any commands send by client will cause "Internal Exception: java.lang.UnsupportedOperationException", and then kicked. Commands send my console is ok.
> 
> This fault has no stacktrace, just "Internal Exception", I dont know how to debug it.
You can attach exception catchers on the netty threads, that'll be usually it unless you're testing with plugins iirc you're not it seems, I'll get on development in your branch too, to help test out and get it ready.

**Comment by PedroMPagani** [comment]
Are you missing something on ProjectileDispenseBehavior? it doesn't build, it's missing .get()

**Comment by ColdeZhang** [comment]
> Are you missing something on ProjectileDispenseBehavior? it doesn't build, it's missing .get()

Yes I leave it alone, because I thought that's upstream's work. It doesnt caused by folia patch, so I keep it out of folia's patch.

**Comment by PedroMPagani** [comment]
Yes, that's a folia change, it should be changed, on paper that's a boolean, on folia for thread safety that's a thread local. IT should be on the patch

**Comment by ColdeZhang** [comment]
> Yes, that's a folia change, it should be changed, on paper that's a boolean, on folia for thread safety that's a thread local. IT should be on the patch

Ok , got it. I may have a missing.

**Comment by ColdeZhang** [comment]
Things should be looks better now.

**Comment by PedroMPagani** [comment]
this all lgtm, seems to be working already tested the majority of the features, including experimental.

File: patches/server/0003-Threaded-Regions.patch [hunk 1]
```diff
@@ -11617,17 +11624,20 @@ index a78ac2c2a536abb0a77371c3d8a9ce0dc5084c38..4f0c962231562c13be8980f98a7d801a
...
 +        // Folia start - region threading
 +        worldserver.loadChunksAsync(
-+            pos, 16, net.minecraft.world.level.chunk.ChunkStatus.FULL,
++            pos, 16, net.minecraft.world.level.chunk.status.ChunkStatus.FULL,
 +            ca.spottedleaf.concurrentutil.executor.standard.PrioritisedExecutor.Priority.NORMAL,
 +            (chunks) -> {
 +                try {
 +        // Folia end - region threading
++
+         PlaceCommand.checkLoaded(worldserver, chunkcoordintpair, chunkcoordintpair);
```
Comment (Spottedleaf) [inline patches/server/0003-Threaded-Regions.patch:11637]: load check should be before chunk load

File: patches/server/0003-Threaded-Regions.patch [hunk 2]
```diff
@@ -13205,6 +13211,7 @@ index 502bdc726b7890b00ee36871d905dea44e8719e3..2f799ead36f0f22afe7dc4b688e59ef2
 -    private final LevelTicks<Fluid> fluidTicks;
 +    //private final LevelTicks<Block> blockTicks; // Folia - region threading
 +    //private final LevelTicks<Fluid> fluidTicks; // Folia - region threading
+     private final PathTypeCache pathTypesByPosCache;
```
Comment (Spottedleaf) [inline patches/server/0003-Threaded-Regions.patch:13214]: did not account for this new field

File: patches/server/0003-Threaded-Regions.patch [hunk 3]
```diff
@@ -15496,43 +15526,45 @@ index 860d33ac56680babd87358b1d7d2216dac3f7865..954809d8997750959541e6462668f17e
...
-         } else {
-             Optional<LastSeenMessages> optional = this.tryHandleChat(packet.lastSeenMessages());
- 
-@@ -2175,21 +2229,22 @@ public class ServerGamePacketListenerImpl extends ServerCommonPacketListenerImpl
+@@ -2168,6 +2220,7 @@ public class ServerGamePacketListenerImpl extends ServerCommonPacketListenerImpl
      @Override
      public void handleChatCommand(ServerboundChatCommandPacket packet) {
-         if (ServerGamePacketListenerImpl.isChatMessageIllegal(packet.command())) {
+         this.tryHandleChat(packet.command(), () -> {
++            this.player.getBukkitEntity().taskScheduler.schedule((ServerPlayer player) -> { // Folia - region threading
```
Comment (Spottedleaf) [inline patches/server/0003-Threaded-Regions.patch:15533]: diff is not needed

File: patches/server/0003-Threaded-Regions.patch [hunk 4]
```diff
@@ -15870,6 +15902,7 @@ index 1e5f709115007ff19901c0a6c3cf884d9e4d3a6c..08daa5ed0b4b599836b97eef1e53e630
 +        ServerLevel worldserver1 = ((CraftWorld)selectedSpawn.getWorld()).getHandle();
 +        player.setPosRaw(selectedSpawn.getX(), selectedSpawn.getY(), selectedSpawn.getZ());
 +        player.lastSave = System.nanoTime(); // changed to nanoTime
++        Optional<CompoundTag> optional = this.load(player); // CraftBukkit - decompile error // Folia - region threading - rewrite login process
```
Comment (Spottedleaf) [inline patches/server/0003-Threaded-Regions.patch:15905]: should not be double-loading the data

File: patches/server/0003-Threaded-Regions.patch [hunk 5]
```diff
@@ -17650,50 +17679,27 @@ index 2439f8d48daca7329049436414f06a36b4b79029..54297b9dde40fe853d873d60373cd8c0
...
-+        if (behaviors.isEmpty()) {
-+            // avoid calling getGameTime, as this may be called while portalling an entity - which will cause
-+            // the world data retrieval to fail
-+            return;
-+        }
-+        // Folia end - region threading
-         long l = entity.level().getGameTime();
- 
--        for(BehaviorControl<? super E> behaviorControl : this.getRunningBehaviors()) {
-+        for(BehaviorControl<? super E> behaviorControl : behaviors) {  // Folia - region threading
```
Comment (Spottedleaf) [inline patches/server/0003-Threaded-Regions.patch:17672]: this diff is needed

File: patches/server/0003-Threaded-Regions.patch [hunk 6]
```diff
@@ -19765,40 +19768,42 @@ index ca89d1593bf1b46c79a882db528cbca1359dc9d4..2c0ce270c947c7fc1d7e0163a4346712
...
+-    private <T> ScheduledTick<T> createTick(BlockPos pos, T type, int delay, TickPriority priority) {
+-        return new ScheduledTick<>(type, pos, this.getLevelData().getGameTime() + (long) delay, priority, this.nextSubTickCount());
 +    // Folia start - region threading
 +    default long getGameTime() {
 +        return this.getLevelData().getGameTime();
-+    }
-+
+     }
+ 
+-    private <T> ScheduledTick<T> createTick(BlockPos pos, T type, int delay) {
```
Comment (Spottedleaf) [inline patches/server/0003-Threaded-Regions.patch:19785]: we shouldn't be changing the visibility unless required

File: patches/server/0014-Do-not-access-POI-data-for-lodestone-compass.patch [hunk 7]
```diff
@@ -1,26 +0,0 @@
...
-diff --git a/src/main/java/net/minecraft/world/item/CompassItem.java b/src/main/java/net/minecraft/world/item/CompassItem.java
-index b3c67c954acf7e518d89d6af65a55d6f22dac059..eea1066379805a4bdef9b496a92ad956d0c84ac7 100644
---- a/src/main/java/net/minecraft/world/item/CompassItem.java
-+++ b/src/main/java/net/minecraft/world/item/CompassItem.java
-@@ -77,7 +77,10 @@ public class CompassItem extends Item implements Vanishable {
-                 Optional<ResourceKey<Level>> optional = getLodestoneDimension(compoundTag);
-                 if (optional.isPresent() && optional.get() == world.dimension() && compoundTag.contains("LodestonePos")) {
-                     BlockPos blockPos = NbtUtils.readBlockPos(compoundTag.getCompound("LodestonePos"));
--                    if (!world.isInWorldBounds(blockPos) || (world.hasChunkAt(blockPos) && !((ServerLevel)world).getPoiManager().existsAtPosition(PoiTypes.LODESTONE, blockPos))) { // Paper - Prevent compass from loading chunks
-+                    // Folia start - do not access the POI data off-region
```
Comment (Spottedleaf) [inline patches/server/0014-Do-not-access-POI-data-for-lodestone-compass.patch:20]: this patch is needed, as the logic was just moved elsewhere

**Comment by Spottedleaf** [comment]
In general we (Paper included) don't take external work for updates, since there's a lot of things involved that can go wrong

https://github.com/PaperMC/Folia/tree/dev/1.20.6

**Pull request closed without merging**